src/appleseed/foundation/core/exceptions/stringexception.h

```diff
@@ -76,7 +76,7 @@ inline StringException::StringException(const char* what)
 
 inline StringException::StringException(const char* what, const char* s)
 {
-    char buf[sizeof(m_what) + sizeof(m_string) + 2];
+    char buf[sizeof(m_what) - 1];
```

Comment:
This causes a "truncation" when using strncpy here:
https://github.com/appleseedhq/appleseed/blob/0acfc054e9167ecfb25e3b5815ed0066e06b43f5/src/appleseed/foundation/core/exceptions/exception.h#L92-L103
`m_what` is going to be assigned to `buf`, and the size of `buf` is unnecessarily larger than `m_what`. Newer versions of gcc require that the size of the source be strictly less than the size of destination, and hence the `-1`

Reply:
This won't work if `m_what` is empty or if `m_what` is smaller that `m_string`. I'm not sure if it's a problem or not, but this need to be confirmed.

Reply:
I'm not sure I get what you mean. "buf" is only assigned to "m_what" so it has to be the same size. If it's any longer, then the rest will be truncated by `strncpy`. Assuming `portable_snprinttf` 
```C++
     portable_snprintf(buf, sizeof(buf), "%s: %s", what, s);
```
acts sensibly assuming and does not crash if size of the formatted string"%s: %s" is larger than `sizeof(buf)` (which it should otherwise why the "n" in s'n'printf) , this should be alright, no?

Reply:
`buf` needs to be big enough to accommodate:

1. The content of `what`
2. The 2-character string `: `
3. The content of `s` (which we accept to truncate to the size of `m_string` if necessary).

This is the reasoning behind making `buf` large enough to accommodate `sizeof(m_what) + sizeof(m_string) + 2` characters (including the terminating `\0`).

Is there a flaw in this reasoning?

Reply:
The reasoning is correct, but in both cases the outcome is the same. Whether `buf` can accommodate "%s : %s" or not, it's going to be truncated anyways to be of size `m_what` as of here
https://github.com/appleseedhq/appleseed/blob/e0bf003c584327e4b4d063e5d2f479f5316a479d/src/appleseed/foundation/core/exceptions/stringexception.h#L79-L81
where we set "m_what = buf". This causes gcc to raise the truncation warning. So the change is to truncate its size from the beginning.